File: lc_go/lcGo/backtrack/generateParenthesis.go
package main

import (
	"fmt"
	"strings"
)

// 生成括号问题 - Go语言实现
// 给定 n 对括号，生成所有有效的括号组合

// 方法1：回溯算法 + 栈验证
func generateParenthesis(n int) []string {
	/*
		🎯 核心思想：回溯 + 栈验证括号有效性
		- 生成所有可能的2n长度序列
		- 用栈验证括号序列是否有效
	*/
	var ans []string
	var path []byte

	var dfs func(i int)
	dfs = func(i int) {
		// 终止条件：生成了2n个字符
		if i == 2*n {
			if isValid(path) {
				ans = append(ans, string(path))
			}
			return
		}

		// 选择 '('
		path = append(path, '(')
		dfs(i + 1)
		path = path[:len(path)-1] // 回溯

		// 选择 ')'
		path = append(path, ')')
		dfs(i + 1)
		path = path[:len(path)-1] // 回溯
	}

	dfs(0)
	return ans
}

// 验证括号序列是否有效
func isValid(brackets []byte) bool {
	var stack []byte
	for _, bracket := range brackets {
		if bracket == '(' {
			stack = append(stack, bracket)
		} else { // bracket == ')'
			if len(stack) == 0 {
				return false
			}
			stack = stack[:len(stack)-1] // 出栈
		}
	}
	return len(stack) == 0
}

// 方法2：优化回溯 - 边构建边验证 (推荐)
func generateParenthesisOptimized(n int) []string {
	/*
		🚀 优化思路：在构建过程中就进行剪枝
		关键观察：
		1. 左括号数量不能超过 n
		2. 右括号数量不能超过左括号数量
		3. 右括号数量不能超过 n
	*/
	var ans []string
	var path []byte

	var dfs func(leftCount, rightCount int)
	dfs = func(leftCount, rightCount int) {
		// 终止条件：生成了 n 对括号
		if leftCount == n && rightCount == n {
			ans = append(ans, string(path))
			return
		}

		// 剪枝 + 选择
		// 选择 '('：左括号数量不能超过 n
		if leftCount < n {
			path = append(path, '(')
			dfs(leftCount+1, rightCount)
			path = path[:len(path)-1] // 回溯
		}

		// 选择 ')'：右括号数量不能超过左括号数量
		if rightCount < leftCount {
			path = append(path, ')')
			dfs(leftCount, rightCount+1)
			path = path[:len(path)-1] // 回溯
		}
	}

	dfs(0, 0)
	return ans
}

// 方法3：字符串拼接版本
func generateParenthesisString(n int) []string {
	/*
		🎨 字符串版本：不用切片，直接拼接字符串
	*/
	var ans []string

	var dfs func(current string, left, right int)
	dfs = func(current string, left, right int) {
		// 终止条件
		if left == n && right == n {
			ans = append(ans, current)
			return
		}

		// 添加左括号
		if left < n {
			dfs(current+"(", left+1, right)
		}

		// 添加右括号
		if right < left {
			dfs(current+")", left, right+1)
		}
	}

	dfs("", 0, 0)
	return ans
}

// 🧪 测试函数
func testGenerateParenthesis() {
	testCases := []struct {
		n        int
		expected []string
	}{
		{1, []string{"()"}},
		{2, []string{"(())", "()()"}},
		{3, []string{"((()))", "(()())", "(())()", "()(())", "()()()"}},
	}

	fmt.Println("=== 生成括号测试 ===")

	for _, tc := range testCases {
		fmt.Printf("\nn = %d (期望 %d 种组合):\n", tc.n, len(tc.expected))

		// 测试方法1：回溯 + 栈验证
		result1 := generateParenthesis(tc.n)
		fmt.Printf("  方法1 (回溯+栈): %v\n", result1)

		// 测试方法2：优化回溯
		result2 := generateParenthesisOptimized(tc.n)
		fmt.Printf("  方法2 (优化回溯): %v\n", result2)

		// 测试方法3：字符串版本
		result3 := generateParenthesisString(tc.n)
		fmt.Printf("  方法3 (字符串版): %v\n", result3)

		// 验证结果数量
		if len(result1) == len(tc.expected) && len(result2) == len(tc.expected) && len(result3) == len(tc.expected) {
			fmt.Printf("  ✅ 所有方法结果数量正确\n")
		} else {
			fmt.Printf("  ❌ 结果数量不正确\n")
		}
	}
}

// 🎨 可视化回溯过程
func visualizeBacktrackProcess(n int) {
	fmt.Printf("\n=== 回溯过程可视化 (n=%d) ===\n", n)

	var ans []string
	var path []byte
	step := 0

	var dfs func(left, right, depth int)
	dfs = func(left, right, depth int) {
		step++
		indent := strings.Repeat("  ", depth)
		fmt.Printf("%s步骤%d: path=%s, left=%d, right=%d\n", indent, step, string(path), left, right)

		// 终止条件
		if left == n && right == n {
			ans = append(ans, string(path))
			fmt.Printf("%s✅ 找到解: %s\n", indent, string(path))
			return
		}

		// 选择 '('
		if left < n {
			fmt.Printf("%s尝试添加 '('\n", indent)
			path = append(path, '(')
			dfs(left+1, right, depth+1)
			path = path[:len(path)-1]
			fmt.Printf("%s回溯，移除 '('\n", indent)
		}

		// 选择 ')'
		if right < left {
			fmt.Printf("%s尝试添加 ')'\n", indent)
			path = append(path, ')')
			dfs(left, right+1, depth+1)
			path = path[:len(path)-1]
			fmt.Printf("%s回溯，移除 ')'\n", indent)
		}
	}

	dfs(0, 0, 0)
	fmt.Printf("\n最终结果: %v\n", ans)
}

func main() {
	// 基本测试
	fmt.Println("🎯 Go语言版本的生成括号算法")
	fmt.Printf("generateParenthesis(3) = %v\n", generateParenthesis(3))
	fmt.Printf("generateParenthesisOptimized(3) = %v\n", generateParenthesisOptimized(3))

	// 运行完整测试
	testGenerateParenthesis()

	// 可视化小例子
	fmt.Println("\n" + strings.Repeat("=", 50))
	visualizeBacktrackProcess(2)

	/*
		📚 Go语言回溯算法学习笔记：

		1. Go的切片特性：
		   - 使用 append() 添加元素
		   - 使用 slice[:len(slice)-1] 移除最后一个元素（回溯）
		   - 切片是引用类型，修改会影响原切片

		2. Go的函数特性：
		   - 使用闭包捕获外部变量
		   - var dfs func(...) 声明递归函数
		   - 函数内部定义函数

		3. Go的字符串处理：
		   - []byte 和 string 的转换
		   - strings.Repeat() 生成重复字符串
		   - 字符串拼接性能考虑

		4. 时间复杂度：
		   - 第 n 个卡特兰数：C_n = (2n)! / ((n+1)! * n!)
		   - 约为 O(4^n / n^(3/2))
	*/
}

File: lc_py/lcPy/backtrack/generateParenthesis.py
'''
生成括号问题 - 经典回溯算法

给定 n 对括号，生成所有有效的括号组合。

示例 1：
输入：n = 3
输出：["((()))","(()())","(())()","()(())","()()()"]

示例 2：
输入：n = 1
输出：["()"]

提示：
1 <= n <= 8
'''
from typing import List


class Solution:
    # 方法1：回溯算法 + 栈验证 (你的思路优化版)
    def generateParenthesis(self, n: int) -> List[str]:
        """
        🎯 核心思想：回溯 + 栈验证括号有效性
        
        你的思路很对！用栈来验证括号的有效性：
        - 遇到 '(' 就入栈
        - 遇到 ')' 就出栈（如果栈为空则无效）
        - 最终栈必须为空
        """
        ans = []
        path = []  # 当前构建的括号序列
        
        def dfs(i: int) -> None:
            # 剪枝：如果当前长度已经超过2n，直接返回
            if i == 2 * n:
                # 验证括号序列是否有效
                if is_valid(path):
                    ans.append(''.join(path))
                return
            
            # 选择 '('
            path.append('(')
            dfs(i + 1)
            path.pop()  # 回溯
            
            # 选择 ')'
            path.append(')')
            dfs(i + 1)
            path.pop()  # 回溯
        
        def is_valid(brackets: List[str]) -> bool:
            """验证括号序列是否有效"""
            stack = []
            for bracket in brackets:
                if bracket == '(':
                    stack.append(bracket)
                else:  # bracket == ')'
                    if not stack:
                        return False
                    stack.pop()
            return len(stack) == 0
        
        dfs(0)
        return ans
    
    # 方法2：优化回溯 - 边构建边验证 (推荐)
    def generateParenthesis_Optimized(self, n: int) -> List[str]:
        """
        🚀 优化思路：在构建过程中就进行剪枝
        
        关键观察：
        1. 左括号数量不能超过 n
        2. 右括号数量不能超过左括号数量
        3. 右括号数量不能超过 n
        """
        ans = []
        path = []
        
        def dfs(left_count: int, right_count: int) -> None:
            # 终止条件：生成了 n 对括号
            if left_count == n and right_count == n:
                ans.append(''.join(path))
                return
            
            # 剪枝 + 选择
            # 选择 '('：左括号数量不能超过 n
            if left_count < n:
                path.append('(')
                dfs(left_count + 1, right_count)
                path.pop()  # 回溯
            
            # 选择 ')'：右括号数量不能超过左括号数量
            if right_count < left_count:
                path.append(')')
                dfs(left_count, right_count + 1)
                path.pop()  # 回溯
        
        dfs(0, 0)
        return ans
    
    # 方法3：字符串拼接版本 (更简洁)
    def generateParenthesis_String(self, n: int) -> List[str]:
        """
        🎨 字符串版本：不用数组，直接拼接字符串
        """
        ans = []
        
        def dfs(current: str, left: int, right: int) -> None:
            # 终止条件
            if left == n and right == n:
                ans.append(current)
                return
            
            # 添加左括号
            if left < n:
                dfs(current + '(', left + 1, right)
            
            # 添加右括号
            if right < left:
                dfs(current + ')', left, right + 1)
        
        dfs('', 0, 0)
        return ans
    
    # 方法4：动态规划解法 (了解即可)
    def generateParenthesis_DP(self, n: int) -> List[str]:
        """
        💡 DP思路：f(n) = "(" + f(i) + ")" + f(n-1-i) for i in [0, n-1]
        
        例如 f(3) 可以分解为：
        - "(" + f(0) + ")" + f(2)
        - "(" + f(1) + ")" + f(1) 
        - "(" + f(2) + ")" + f(0)
        """
        if n == 0:
            return ['']
        
        ans = []
        for i in range(n):
            # 左边 i 对括号，右边 n-1-i 对括号
            for left in self.generateParenthesis_DP(i):
                for right in self.generateParenthesis_DP(n - 1 - i):
                    ans.append(f'({left}){right}')
        
        return ans


# 📚 回溯算法学习笔记
"""
🎯 回溯算法核心思想：

1. 什么是回溯？
   - 尝试所有可能的解决方案
   - 当发现当前路径不可行时，"回退"到上一步
   - 继续尝试其他可能性

2. 回溯三要素：
   - 选择：每一步有哪些选择？(本题：选 '(' 或 ')')
   - 约束：什么情况下选择无效？(本题：右括号不能超过左括号)
   - 目标：什么时候找到解？(本题：生成了 n 对括号)

3. 回溯模板：
   def backtrack(路径, 选择列表):
       if 满足结束条件:
           result.add(路径)
           return
       
       for 选择 in 选择列表:
           做选择
           backtrack(路径, 选择列表)
           撤销选择  # 这就是"回溯"

4. 本题的关键剪枝：
   - 左括号数量 <= n
   - 右括号数量 <= 左括号数量
   - 这样可以避免生成无效的括号序列

5. 时间复杂度：
   - 第 n 个卡特兰数：C_n = (2n)! / ((n+1)! * n!)
   - 约为 O(4^n / n^(3/2))
"""


# 🧪 测试函数
def test_generate_parenthesis():
    solution = Solution()
    
    test_cases = [
        (1, ["()"]),
        (2, ["(())", "()()"]),
        (3, ["((()))", "(()())", "(())()", "()(())", "()()()"])
    ]
    
    print("=== 生成括号测试 ===")
    
    for n, expected in test_cases:
        print(f"\nn = {n} (期望 {len(expected)} 种组合):")
        
        # 测试方法1：回溯 + 栈验证
        result1 = solution.generateParenthesis(n)
        result1.sort()
        print(f"  方法1 (回溯+栈): {result1}")
        
        # 测试方法2：优化回溯
        result2 = solution.generateParenthesis_Optimized(n)
        result2.sort()
        print(f"  方法2 (优化回溯): {result2}")
        
        # 测试方法3：字符串版本
        result3 = solution.generateParenthesis_String(n)
        result3.sort()
        print(f"  方法3 (字符串版): {result3}")
        
        # 验证结果
        expected_sorted = sorted(expected)
        if result1 == result2 == result3 == expected_sorted:
            print(f"  ✅ 所有方法结果一致且正确")
        else:
            print(f"  ❌ 结果不一致或错误")
            print(f"     期望: {expected_sorted}")


# 🎨 可视化回溯过程
def visualize_backtrack_process(n: int):
    print(f"\n=== 回溯过程可视化 (n={n}) ===")
    
    ans = []
    path = []
    step = [0]  # 用列表包装以便在内部函数中修改
    
    def dfs(left: int, right: int, depth: int = 0) -> None:
        step[0] += 1
        indent = "  " * depth
        print(f"{indent}步骤{step[0]}: path={''.join(path)}, left={left}, right={right}")
        
        # 终止条件
        if left == n and right == n:
            ans.append(''.join(path))
            print(f"{indent}✅ 找到解: {''.join(path)}")
            return
        
        # 选择 '('
        if left < n:
            print(f"{indent}尝试添加 '('")
            path.append('(')
            dfs(left + 1, right, depth + 1)
            path.pop()
            print(f"{indent}回溯，移除 '('")
        
        # 选择 ')'
        if right < left:
            print(f"{indent}尝试添加 ')'")
            path.append(')')
            dfs(left, right + 1, depth + 1)
            path.pop()
            print(f"{indent}回溯，移除 ')'")
    
    dfs(0, 0)
    print(f"\n最终结果: {ans}")
    return ans


if __name__ == '__main__':
    test_generate_parenthesis()
    
    # 可视化小例子
    print("\n" + "="*50)
    visualize_backtrack_process(2)